lib/pbench/server/cache_manager.py

```diff
@@ -261,68 +281,77 @@ def create(cls, tarball: Path, controller: "Controller") -> "Tarball":
 
         return cls(destination, controller)
 
-    def cache_map(self, tar_path: Path):
+    def cache_map(self, dir_path: Path):
         """Builds Hierarchy structure of a Directory in a Dictionary
         Format.
 
         Args:
-            dir_path: root of a directory
+            dir_path: root directory
         """
-        cmap = {tar_path.name: {"details": FileInfo(tar_path)}}
-        dir_queue = deque([(tar_path, cmap)])
+        cmap = {dir_path.name: {"details": FileInfo(self.name, dir_path)}}
+        dir_queue = deque([(dir_path, cmap)])
         while dir_queue:
-            tar_path, parent_map = dir_queue.popleft()
-            tar_n = tar_path.name
+            dir_path, parent_map = dir_queue.popleft()
+            tar_n = dir_path.name
             curr = parent_map[tar_n]
 
-            dir_list = list(tar_path.glob("*"))
+            dir_list = list(dir_path.glob("*"))
             if dir_list:
                 curr["children"] = {}
                 curr = curr["children"]
 
             for l_path in dir_list:
-                tar_info = FileInfo(l_path)
-                curr[l_path.name] = {"details": tar_info}
+                dir_info = FileInfo(self.name, l_path)
+                curr[l_path.name] = {"details": dir_info}
                 if l_path.is_dir():
                     dir_queue.append((l_path, curr))
         self.cachemap = cmap
 
     @staticmethod
-    def traverse_cmap(path: Path, cachemap: dict) -> dict:
+    def traverse_cmap(path: Path, cachemap: dict) -> dict[str, dict]:
+        """Sequentially traverses the cachemap to find the leaf of a
+        relative path reference
+
+        Args:
+            path: relative path of the sub-directory/file
+            cachemap: dictionary mapping of the root Dicrectory
+
+        Returns:
+            Dictionary with directory/file details or children if present
+        """
         name = path.name
         file_list = path.parts
         c_map = cachemap
-        i = 0
-
-        if file_list[i] in ("/", "//"):
-            i += 1
 
-        while i < len(file_list):
-            c_map = c_map[file_list[i]]
+        for file_l in file_list:
+            c_map = c_map[file_l]
             if c_map["details"].type == "directory":
-                if "children" in c_map and file_list[i] != name:
+                if "children" in c_map and file_l != name:
                     c_map = c_map["children"]
-            i += 1
 
         return c_map
```

Comment:
> One of the issues with this code is that it doesn't quite follow the structure of the data that it is traversing: at the top level of the cachemap, `c_map` is a `dict` with (one) keys corresponding to the file entries (i.e., it's a "directory"); at the next level, there's a `dict` with keys `"details"` and `"children"` and _not_ file entries; then, at the next level, we're back to file entries again.

The structure of the data is quite consistent. If there are children(sub-directory or file) in a directory, that there will be a `children` and `details` key else this children key is ommited. If there is no children key than there is only `details` key under that directory.
```
cachemap =  {'dir_name': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c3590>, 'children': {'subdir1': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c3890>, 'children': {'subdir13': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c1d10>, 'children': {'subdir111': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c3c10>, 'children': {'f1112_sym': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c2b50>}, 'f1111.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c0950>}}}}}, 'f11.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c1210>}, 'subdir12': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c20d0>}, 'subdir11': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c3ad0>}}}, 'f1.json': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fc2143c25d0>}}}}
```

This is the same structure as displayed in tests:

```
Generated cache map

            {
                'dir_name': {
                    'details': <cache_manager.FileInfo object>,
                    'children': {
                        'f1.json': {'details': <cache_manager.FileInfo object>},
                        'subdir1': {
                            'details': <cache_manager.FileInfo object>,
                            'children': {
                                'f11.txt': {'details': <cache_manager.FileInfo object>},
                                'subdir13': {
                                    'details': <cache_manager.FileInfo object>,
                                    'children': {
                                        'subdir111': {
                                            'details': <cache_manager.FileInfo object>,
                                            'children':{
                                                'f1112_sym': { 'details': <cache_manager.FileInfo object>}
                                                'f1111.txt': { 'details': <cache_manager.FileInfo object>}
                                            }}}},
                                'subdir12': {'details': <cache_manager.FileInfo object>},
                                'subdir11': {'details': <cache_manager.FileInfo object at>
                                }}}}}}
```

Reply:
> From there, I would (again) point out that an `info` always has a `"children"` key (right??)

This is only possible if there are sub-directories in a directory. If there are no sub-directories or files there will be no `children` key. I have added this empty `children` key ealier but was asked to remove if there are no entries there.

Reply:
> `name` _could_ appear as both a subdirectory and a leaf in the same path, we shouldn't be checking for it in the middle of the loop.

This is a case when the given path is a directory but we don't want to go to the "children" if the last sub-dir because then we miss out on details of the directory.

This particular issue, maybe I can demonstrate the need of it with two example:

This is an example with checking last directory name:

So, as soon as it checked that the name of the sub-directory is similar to the last entry of the path it just returns the value inside that directory and not the value of `children` key.

```
path ==  dir_name/subdir1
c_map ==  {'dir_name': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051550ed0>, 'children': {'subdir1': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125150>, 'children': {'subdir13': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125250>, 'children': {'subdir111': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125590>, 'children': {'f1112_sym': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125a50>}, 'f1111.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125c50>}}}}}, 'f11.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125410>}, 'subdir12': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125690>}, 'subdir11': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125710>}}}, 'f1.json': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe0511252d0>}}}}
name ==  subdir1
c_map ==  {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125150>, 'children': {'subdir13': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125250>, 'children': {'subdir111': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125590>, 'children': {'f1112_sym': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125a50>}, 'f1111.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125c50>}}}}}, 'f11.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125410>}, 'subdir12': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125690>}, 'subdir11': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fe051125710>}}}
```

This is without checking last directory name:

In this it is returning the value of  the `children` key of the last dir in path and we are missing out on `details` of the last directory.

```
path ==  dir_name/subdir1
c_map ==  {'dir_name': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dced0>, 'children': {'subdir1': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd050>, 'children': {'subdir13': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd150>, 'children': {'subdir111': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd490>, 'children': {'f1112_sym': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd950>}, 'f1111.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337ddb50>}}}}}, 'f11.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd310>}, 'subdir12': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd590>}, 'subdir11': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd610>}}}, 'f1.json': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd1d0>}}}}
name ==  subdir1
c_map ==  {'subdir13': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd150>, 'children': {'subdir111': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd490>, 'children': {'f1112_sym': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd950>}, 'f1111.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337ddb50>}}}}}, 'f11.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd310>}, 'subdir12': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd590>}, 'subdir11': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fee337dd610>}}
```

Reply:
```        # omit the `name` local variable -- we only use it once, now
        file_list = path.parts[:-1]
        f_entries = cachemap

        for file_l in file_list:
            info = f_entries[file_l]
            if info["details"].type == "directory":
                f_entries = info["children"]

        return f_entries[path.name]
```
This is good if we want to omit last last directory/file in path. but we want to go till the end. If we don't go until the end then we will need to do extra processing in `get_info`.

following will be the returned value for path "dir_name/f1.json" and we will need to go through each key to find the last directory/file.

```
path ==  dir_name/f1.json
c_map ==  {'subdir1': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e4d50>, 'children': {'subdir13': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e4e50>, 'children': {'subdir111': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e5190>, 'children': {'f1112_sym': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e5650>}, 'f1111.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e5850>}}}}}, 'f11.txt': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e5010>}, 'subdir12': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e5290>}, 'subdir11': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e5310>}}}, 'f1.json': {'details': <pbench.server.cache_manager.FileInfo object at 0x7fd0ce3e4ed0>}}
```

Reply:
After the addition of `path.name` in `f_entries[path.name]` it does seem to have aligned with our code.

Reply:
> There's one weak point which is if `info` doesn't have `"details"` or `"children"` then we'll find our way to the `except` block and raise a confusing message, but let's resolve that by making sure that `info` is always properly constructed.

T believe that this is not a weak but validpoint. A second last element of the path should be a directory and not a file. if the second last element doesn't have `children` that the path is wrong and we are supposed to raise and exception about that.

Reply:
> The structure of the data is quite consistent.

I agree.  My point about inconsistency was around using the same variable to refer to the values of a file/directory key and the values of the `"children"` key.  I think that using separate variables for those two different kinds of values makes things much clearer in the code.

> If there are no sub-directories or files there will be no children key. I have added this empty children key earlier but was asked to remove if there are no entries there. Due to the same reason we will have to check for it.

I understand that, now.  I couldn't find where you were asked to remove the `"children"` key; nevertheless, I'd like to recommend that (for "directory"-type files) the `"details"` and `"children"` keys always be present (modulo what you're working on now for a `@dataclass` hierarchy 🙂):  if we know from context that we have a cache entry for a "directory", we shouldn't need to check whether there is a `"children"` key present -- that leads to cumbersome code, and it's nicer to be able to just reference the list (and, if it's empty, that's just as meaningful as if the key were absent).

>> name could appear as both a subdirectory and a leaf in the same path, we shouldn't be checking for it in the middle of the loop.
>
> This is a case when the given path is a directory but we don't want to go to the "children" if the last sub-dir because then we miss out on details of the directory.

Looking at the latest revision of your code, hopefully this is moot now.  (I'm hoping that adding this case to the tests proved useful.  🙂)

> This is good if we want to omit last last directory/file in path. but we want to go till the end. If we don't go until the end then we will need to do extra processing in get_info.

Again, looking at the latest revision of your code, hopefully, this is moot, too.

>> There's one weak point which is if info doesn't have "details" or "children" then we'll find our way to the except block and raise a confusing message, but let's resolve that by making sure that info is always properly constructed.
>
> I believe that this is not a weak but validpoint. A second last element of the path should be a directory and not a file or it is not a valid path. if the second last element doesn't have children that the path is wrong and we are supposed to raise an exception about that.

Happily, you added a test case which shows this problem, and I called it out in [a comment](https://github.com/distributed-system-analysis/pbench/pull/3396#discussion_r1191464513) below.  The problem is that we have two different types of dictionaries, and errors referencing them should be handled differently from each other (but, as the code stands at the moment, there is only a single `try` block and a single response to the error).  If we are looking up a file in a "children" dictionary and cannot find it, we need to report that that path element is not found.  However, if we are looking up the `"children"` key in an "info" dictionary, then we need to report that the current directory is unexpectedly empty, not that "File/directory _children_ in path {path} not found in cache."  (And, if the key is `"details"`, then something terrible has happened....)

I think we can solve this problem in one of two ways.  Either we can explicitly check for the `"children"` key (presumably by using `get()` to fetch the value and then check the result for `None`); or, we can make sure that the `"children"` key always exists, pointing at an empty list if necessary.  Doing the former allows us to raise a customized exception message for it; doing the latter allows us to handle the situation with the existing file-not-found code.  (As I've said, I vote for the latter.  😉)

Reply:
The later one sounds good to me as well.